snf-image-helper/tasks/20FilesystemResizeUnmounted.in

```diff
@@ -74,11 +74,16 @@ if [[ "$ptype" == ext[234] ]]; then
     state=$($TUNE2FS -l "$device" | grep ^Filesystem\ state: | cut -d: -f2);
     state=$(echo $state) #trim the value
 
-    # We force a file system resize here if the file system is clean, even if
-    # resize2fs complains. By default resize2fs will refuse to resize a file
-    # system that has been mounted after the last fs check, but since we are
-    # sure the file system is clean it's safe enough to bypass this.
+    # NOTE: A filesystem might be clean but its journal may need recovery. This
+    # is typical when the filesystem is a snapshot of a running VM. So we do an
+    # fsck, which will fix the journal and then we resize the filesystem,
+    # otherwise we almost always end up with a corrupted filesystem and an
+    # un-bootable VM.
     if [ "$state" = "clean" ]; then
+        # NOTE: We use automatic repair (-p) because otherwise e2fsck complains
+        # that it needs a terminal for interactive repairs.
+        $E2FSCK -p "$device" || \
```

Comment:
This is from the man page of `e2fsck`:
```
EXIT CODE
       The exit code returned by e2fsck is the sum  of  the  following  condi‐
       tions:
            0    - No errors
            1    - File system errors corrected
            2    - File system errors corrected, system should
                   be rebooted
            4    - File system errors left uncorrected
            8    - Operational error
            16   - Usage or syntax error
            32   - E2fsck canceled by user request
            128  - Shared library error
```
I haven't tried it but If the journal gets recovered by `e2fsck`, I assume the utility would return `1`. I think that if the output is 1 or 2, which means that all error are corrected, snf-image shouldn't fail. Something like this makes more sense to me:

```
current_options=$(set +o)
set +e
$E2FSCK -p "$device"
rc=$?
if [ $rc -gt 2 ]; then
    log_error "Filesystem was clean but fsck exited with error code $?"
fi
eval "$current_options"
 ```
What do you think?